File: matlab/generate_RD_curves_base.m
function generate_RD_curves_base(archname,tranname,testsize,inlayers,outlayer)

% Choose one of: 'alexnet', 'vgg16', 'densenet201', 'mobilenetv2' and
% 'resnet50', and specify the filepath for ILSVRC test images. Number
% of test files to predict can be set manually or set to 0 to predict
% all files in the datastore (not recommended)
% archname = 'alexnet';
imagedir = '~/Developer/ILSVRC2012_val/*.JPEG';
labeldir = './ILSVRC2012_val.txt';
% tranname = 'idt2';
% testsize = 1024;
maxsteps = 32;
maxrates = 17;

[neural,images] = loadnetwork(archname,imagedir, labeldir, testsize);
[layers,lclass] = removeLastLayer(neural);
neural = assembleNetwork(layers);
nclass = assembleNetwork(lclass);
trans = {str2func(tranname), str2func(['i',tranname])};

l_kernel = findconv(neural.Layers);
l_length = length(l_kernel);

hist_delta = cell(l_length,1);
hist_coded = cell(l_length,1);
hist_W_sse = cell(l_length,1);
hist_Y_sse = cell(l_length,1);
hist_Y_top = cell(l_length,1);

[Y,Y_cats] = pred(neural,nclass,images,outlayer);
disp(sprintf('%s | top1: %4.1f', archname, 100*mean(images.Labels == Y_cats)));

layers = neural.Layers(l_kernel);
for l = inlayers
    layer = layers(l);
    K = gettrans(tranname,archname,l);
    [h,w,p,q] = size(layer.Weights);
    layer.Weights = transform(layer.Weights,K{1});
    hist_delta{l} = zeros(maxrates,maxsteps,h*w,1)*NaN;
    hist_coded{l} = zeros(maxrates,maxsteps,h*w,1)*NaN;
    hist_W_sse{l} = zeros(maxrates,maxsteps,h*w,1)*NaN;
    hist_Y_sse{l} = zeros(maxrates,maxsteps,h*w,testsize)*NaN;
    hist_Y_top{l} = zeros(maxrates,maxsteps,h*w,testsize)*NaN;
    
    for i = 1 % treat all frequencies or bands as one

        scale = floor(log2(sqrt(mean(layer.Weights(:).^2))));
        coded = Inf;
        offset = scale + 2;
        for k = 1:maxrates %number of bits
            B = k - 1;
            last_Y_sse = Inf;
            last_W_sse = Inf;
            for j = 1:maxsteps
                % quantize each of the q slices
                quant = layer;
                delta = offset + 0.25*(j-1);
                quant.Weights(:) = quantize(quant.Weights(:),2^delta,B);
                coded = B*(p*q); %qentropy(quant.Weights(:),B)*(h*w*p*q);
                % assemble the net using layers
                quant.Weights = transform(quant.Weights,K{2});
                ournet = replaceLayers(neural,quant);

                [Y_hats,Y_cats] = pred(ournet,nclass,images,outlayer);
                hist_Y_sse{l}(k,j,:,:) = repmat(mean((Y_hats - Y).^2)*(1/(h*w)),[h*w,1]);
                hist_Y_top{l}(k,j,:,:) = repmat((images.Labels == Y_cats)',[h*w,1]);
                hist_W_sse{l}(k,j,:,1) = mean((quant.Weights(:) - neural.Layers(l_kernel(l)).Weights(:)).^2)*(1/(h*w));
                hist_delta{l}(k,j,:,1) = delta;
                hist_coded{l}(k,j,:,1) = coded;
                mean_Y_sse = mean(hist_Y_sse{l}(k,j,i,:));
                mean_W_sse = mean(hist_W_sse{l}(k,j,i,1));
                disp(sprintf('%s %s | layer: %03d/%03d, band: all/%03d, scale: %3d, delta: %+6.2f, ymse: %5.2e, wmse: %5.2e, top1: %4.1f, rate: %5.2e', ...
                             archname, tranname, l, l_length, h*w, scale, delta, mean_Y_sse, ...
                             mean_W_sse, 100*mean(hist_Y_top{l}(k,j,i,:)), coded/(p*q)));
                if (mean_Y_sse > last_Y_sse) && ...
                   (mean_W_sse > last_W_sse) || ...
                   (B == 0)     
                    [~,j] = min(mean(hist_Y_sse{l}(k,:,i,:),4));
                    delta = hist_delta{l}(k,j,i,1);
                    offset = delta - 2;
                    break;
                end
                last_Y_sse = mean_Y_sse;
                last_W_sse = mean_W_sse;
            end
        end
    end
end
save(sprintf('%s_%s_base_val_%d_%s',archname,tranname,testsize,outlayer),'hist_coded','hist_Y_sse','hist_Y_top','hist_delta','hist_W_sse');

File: python/generate_RD_curves_base.py
import common
import header
import importlib
importlib.reload(common)
importlib.reload(header)

from common import *
from header import *

trantype = str(sys.argv[1])
tranname = str(sys.argv[2])
archname = str(sys.argv[3])
testsize = int(sys.argv[4])

maxsteps = 32
maxrates = 17

neural, images, labels = loadnetwork(archname,gpuid,testsize)

neural.eval()
Y = predict(neural,images)
Y_cats = gettop1(Y)
mean_Y_top = (Y_cats == labels).double().mean()
print('%s %s | top1: %5.2f' % (archname, tranname, 100*mean_Y_top))

perm, flip = getperm(trantype)
layers = findconv(neural,False)

for l in range(0,len(layers)):
    with torch.no_grad():
        basis_vectors = gettrans(archname,trantype,tranname,l,'')
        layer_weights = layers[l].weight.clone()
        layer_weights = layer_weights.flatten(2).permute(perm)
        dimen_weights = layer_weights.size()
        layer_weights = layer_weights.flatten(1).permute(flip)
        trans_weights = basis_vectors[:,:,0].flatten(1).mm(layer_weights)
        basis_vectors = basis_vectors[:,:,1].flatten(1)
        base_delta = torch.ones(maxrates,maxsteps,trans_weights.size(0),device=getdevice()) * Inf
        base_coded = torch.ones(maxrates,maxsteps,trans_weights.size(0),device=getdevice()) * Inf
        base_W_sse = torch.ones(maxrates,maxsteps,trans_weights.size(0),device=getdevice()) * Inf
        base_Y_sse = torch.ones(maxrates,maxsteps,trans_weights.size(0),device=getdevice()) * Inf
        base_Y_top = torch.ones(maxrates,maxsteps,trans_weights.size(0),device=getdevice()) * Inf
        s = min(int(np.ceil(trans_weights.size(0)/8)),int(np.ceil(trans_weights.size(1)/8)))

        for i in range(0,trans_weights.shape[0],s):
            rs = range(i,min(i+s,trans_weights.shape[0]))
            scale = (trans_weights[rs,:].reshape(-1)**2).mean().sqrt().log2().floor()
            if scale < -24:
                continue
            scale = (basis_vectors[:,rs].reshape(-1)**2).mean().sqrt().log2().floor()

            coded = Inf
            start = scale - 2
            for b in range(0,maxrates):
                last_Y_sse = Inf
                last_W_sse = Inf
                for j in range(0,maxsteps):
                    sec = time.time()
                    delta = start + 0.25*j
                    coded = s*basis_vectors.shape[0]
                    quant_vectors = basis_vectors.clone()
                    quant_vectors[:,rs] = quantize(quant_vectors[:,rs],2**delta,b)
                    delta_weights = (quant_vectors[:,rs]-basis_vectors[:,rs]).mm(trans_weights[rs,:])
                    layers[l].weight[:] = (layer_weights + delta_weights).permute(inv(flip)).\
                                          reshape(dimen_weights).permute(inv(perm)).reshape(layers[l].weight.shape)
                    Y_hats = predict(neural,images)
                    Y_cats = gettop1(Y_hats)
                    Y_exps = Y_hats.exp()/(Y_hats.exp().sum(1)).reshape(-1,1)
                    sec = time.time() - sec
                    base_W_sse[b,j,i] = (delta_weights**2).mean()
                    #base_Y_log[b,j,i] = -Y_exps[:,0:1000:1000//testsize].diag().log().sum()
                    base_Y_sse[b,j,i] = ((Y_hats - Y)**2).mean()
                    base_Y_top[b,j,i] = (Y_cats == labels).double().mean()
                    base_delta[b,j,i] = delta
                    base_coded[b,j,i] = coded*b
                    mean_Y_sse = base_Y_sse[b,j,i]
                    mean_Y_top = base_Y_top[b,j,i]
                    mean_W_sse = base_W_sse[b,j,i]
                    mean_coded = base_coded[b,j,i]

                    if mean_Y_sse > last_Y_sse and \
                       mean_W_sse > last_W_sse or  \
                       b == 0:
                        break

                    last_Y_sse = mean_Y_sse
                    last_W_sse = mean_W_sse

                _,  j = base_Y_sse[b,:,i].min(0)
                delta = base_delta[b,j,i]
                start = delta - 2
                mean_Y_sse = base_Y_sse[b,j,i]
                mean_Y_top = base_Y_top[b,j,i]
                print('%s %s | layer: %03d/%03d, band %04d/%04d, delta: %+6.2f, '
                      'mse: %5.2e (%5.2e), top1: %5.2f, rate: %4.1f, time: %5.2fs'\
                      % (archname, tranname, l, len(layers), i, basis_vectors.shape[1],\
                         delta, mean_Y_sse, mean_W_sse, 100*mean_Y_top, b, sec))

        layers[l].weight[:] = layer_weights[:].permute(inv(flip)).reshape(dimen_weights).\
                              permute(inv(perm)).reshape(layers[l].weight.shape)
        io.savemat(('%s_%s_val_%03d_%04d_output_%s_base.mat' % (archname,tranname,l,testsize,trantype)),\
                   {'base_coded':base_coded.cpu().numpy(),'base_Y_sse':base_Y_sse.cpu().numpy(),\
                    'base_Y_top':base_Y_top.cpu().numpy(),'base_delta':base_delta.cpu().numpy(),\
                    'base_W_sse':base_W_sse.cpu().numpy()})
